### ISSUE TITLE
IMG: Allow more than 64 sectors per track

### ISSUE THREAD

**Issue by Simone2018** [keirf/FlashFloppy#211]
It's Amazing... I tested succesfully 16Mb Floppy Disk with it!
Can you unlock maximum number of sectors from 64 to 128 ?
I have reached 32768 sectors... the floppy FAT12 BPB has a maximum of FFFFh sectors so I want to reach the theorical maximum (32Mb floppy!)
Thank you

**Comment by keirf** [comment]
Is there scope for increasing physical sector size to say 4kb? If your bios can handle more sectors per track could it also handle larger sectors and grow space that way?

**Comment by Simone2018** [comment]
I'll try larger Sector size but i.e. Winimage support 512Kb sectors only. And I suppose MSDOS, IO.SYS etc etc all support 512Kb Sector only... so increasing sectors (and maybe sides more than 2) could be win-win. 32Mb is the FAT12 structural limit (but MBR could support more than that). If it's too difficult don't lose too much time.

**Comment by keirf** [comment]
Increasing sides would be a bit tricky as there is only one head-select pin ;)

**Comment by keirf** [comment]
Increasing max sectors yes not really that hard. Surprised DOS is flexible on that but not physical sector size? I suppose old hard drives had 512b sectors too?

**Comment by Simone2018** [comment]
Yes, all (floppy, Hard Disk) was based on 512b sectors... buffers, minial data read, boot sequence, MBR, etc etc. Maybe old dos (<2.1) was less fixed on this.
Sector number is flexible because Hard Disk geometry was using it. To allow as much space I had to install 2M-XBIOS for floppy (it support also HD floppy but not so my controller)...

**Comment by keirf** [comment]
I could even increase the limit to 256 sectors, I suppose.

**Comment by Simone2018** [comment]
Talking for IBM-PC view:
"for disk access with the BIOS of IBM-PC compatible machines, the sector number was encoded in six bits, resulting in a maximal number of 111111 (63) sectors per track"
but 256 is welcome! :) thank you

**Comment by keirf** [comment]
If there's a maximum support of 64 sectors in the IBM-PC BIOS, why is there benefit to increasing at all beyond the current 64?

**Comment by Simone2018** [comment]
only for NON-IBM i think; 
but I can also write a new INT13h function to take it in account...

**Comment by tlo2357** [comment]
The WD 179x controllers have 8-bit registers for the track and sector numbers.  I haven't seen anything in the datasheet stating that the full range isn't available.

I expect that reading unusual track or sector numbers in the ID field using a 179x should work but writing some of them (during format) might produce incorrect results because they might be interpreted as format control bytes, but a drive emulator shouldn't care about that.  If I can find the time and it hasn't already been done somewhere it *would* be interesting to test and document this with my Z-100 which uses a 1797.

**Comment by keirf** [comment]
This is in the master branch now (pending v2.1a)

**Comment by tlo2357** [comment]
Some may be curious about the test results.  PC controllers *do* support track and sector numbers ranging from 0 through 255, although I suppose some wonky Super I/O chips might not – similar to how they don't all support 128-byte sectors.

I ran tests on my old test PC which has an ITE IT8712F-S Super I/O chip, which Linux identifies as a "post-1991 82077" controller.  I used fdrawcmd to format track zero of a real disk with different track and sector ID number combinations to test what would be accepted by the controller's read sector command.  These tests showed that all track and sector values worked.

To spot-check that the controller was actually writing to disk what it was supposed to, not truncating anything, I crafted a test and read the MFM data from the track using a Raspberry Pi.  All of the track and sector ID values were as expected, including 0 and 255.
